src/converters/app-converter.ts

```diff
@@ -15,13 +15,125 @@ import {
+export interface SecretsManagerDynamicReferenceEnd {
+    /**
+     * Currently, the only supported value is SecretString. The default is SecretString.
+     */
+    secretString?: string;
+
+    /**
+     * The key name of the key-value pair whose value you want to retrieve.
+     * If you don't specify a json-key, CloudFormation retrieves the entire secret text.
+     *
+     * This segment may not include the colon character ( :).
+     */
+    jsonKey?: string;
+
+    /**
+     * The staging label of the version of the secret to use.
+     * If you use version-stage then don't specify version-id.
+     * If you don't specify either version-stage or version-id, then the default is the AWSCURRENT version.
+     *
+     * This segment may not include the colon character ( :).
+     */
+    versionStage?: string;
+
+    /**
+     * The unique identifier of the version of the secret to use.
+     * If you specify version-id, then don't specify version-stage.
+     * If you don't specify either version-stage or version-id, then the default is the AWSCURRENT version.
+     *
+     * This segment may not include the colon character ( :).
+     */
+    versionId?: string;
+}
+export interface SecretsManagerDynamicReference extends SecretsManagerDynamicReferenceEnd {
+    /**
+     * The name or ARN of the secret.
+     *
+     * To access a secret in your AWS account, you need only specify the secret name.
+     * To access a secret in a different AWS account, specify the complete ARN of the secret.
+     */
+    secretId: string;
+}
+
+/**
+ * Parses a secretsmanager dynamic reference into its components. This function should be used to resolve
+ * references that are complete strings, i.e. no unresolved values
+ *
+ * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/dynamic-references-secretsmanager.html
+ *
+ * @param secret - The secretsmanager dynamic reference, i.e. {{resolve:secretsmanager:secret-id:secret-string:json-key:version-stage:version-id}}
+ * @returns the matched secret reference
+ */
+export function parseDynamicSecretReference(secret: string): SecretsManagerDynamicReference {
+    const match = secret.match(SECRETS_MANAGER_DYNAMIC_REGEX);
+    if (match) {
+        const [_, secretId, secretString, jsonKey, versionStage, versionId] = match;
+        return {
+            secretId,
+            secretString: secretString || undefined,
+            jsonKey: jsonKey || undefined,
+            versionStage: versionStage || undefined,
+            versionId: versionId || undefined,
+        };
+    }
+    throw new Error(`Invalid Secrets Manager dynamic reference: value: ${secret}`);
+}
+
+/**
+ * Parses a secretsmanager dynamic reference into its components. This function should be used to resolve
+ * the end part of the reference string from the :secret-string onwards
+ *
+ * This is useful in cases where the full reference contains unresolved values, e.g.
+ *
+ *   "MasterUserPassword": {
```

Comment:
This is really really strange, the execution model here is strange.  I have been looking up some references here. Sounds like https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/dynamic-references-secretsmanager.html is describing the feature, and AI gave me this example:

```
Resources:
  MySecret:
    Type: AWS::SecretsManager::Secret
    Properties:
      Name: MyDatabaseSecret
      Description: "A secret for my database credentials"
      SecretString: !Sub |
        {
          "username": "${DBUsername}",
          "password": "${DBPassword}"
        }

  MyDatabase:
    Type: AWS::RDS::DBInstance
    Properties:
      DBInstanceIdentifier: MyDBInstance
      MasterUsername: !Join
        - ""
        - - '{{resolve:secretsmanager:'
          - !Ref MySecret
          - ':SecretString:username}}'
      MasterUserPassword: !Join
        - ""
        - - '{{resolve:secretsmanager:'
          - !Ref MySecret
          - ':SecretString:password}}'

```

I would guess that CF takes this form and:

1. first, resolves Ref references
2. second, computes Fn::Join heuristic to get a string
3. finally, searches the string for dynamic references and substitutes them

Is that right? 

If we're parsing Fn::Join form this means we have a very different evaluation order that will work for Fn::Join but fail to compose the same way if other intrinsics are in play. So this is a very partial solution.  Is getting to emulate the CF evaluation sequence not feasible or is something we can backlog?

Reply:
Ok, i've updated this to (I think) do what you are saying.